### ISSUE TITLE
`git lfs prune` hangs

### ISSUE THREAD

**Issue by pwagland** [git-lfs/git-lfs#3403]
First up, I know that this looks like a duplicate of #1494, #3095, or #1860. However, the first is a really old issue, wherein it was asked to create a new issue to reproduce, which happened in #3095. This in turn was closed since the submitter could no longer reproduce, and #1860 was resolved long ago, and I am seeing the issue in:
```
12:14 $ git lfs env
git-lfs/2.6.0 (GitHub; darwin amd64; go 1.11.2)
git version 2.19.2
```

Each time that I run `GIT_TRACE=2 git lfs prune`, it hangs with the same last lines of output (obviously with different timestamps!):

```
12:07:39.385073 trace git-lfs: RETAIN: 11625cc4f64165de32eba059e310a7b7caf89dfa3e4c34432a222833320d9208 unpushed
12:07:39.385079 trace git-lfs: RETAIN: 2c0048ab3ce75a202f692b159d6aa0a68edce3e4e4c5123a3359a38b29faa6b1 unpushed
12:07:39.385088 trace git-lfs: RETAIN: d29825317cfc2a0316c2d3182222f223e55a44a193c2738f607b19af8930a715 unpushed
12:07:39.385134 trace git-lfs: RETAIN: 6108a2e9db3a84fba7875339948230e99f62bb674fc2baa2c1e24717148c7e27 unpushed
12:07:39.385164 trace git-lfs: RETAIN: 11625cc4f64165de32eba059e310a7b7caf89dfa3e4c34432a222833320d9208 unpushed
12:07:39.414052 trace git-lfs: RETAIN: 2c0048ab3ce75a202f692b159d6aa0a68edce3e4e4c5123a3359a38b29faa6b1 unpushed
```

At this point it just seems to hang. When I do a `top`, there is no CPU, nor disk activity. I have roughly 100 git processes running, and they never change, the processes that I have are:

```
12:19 $ ps -ef | grep git | cut -c 55- | sort | uniq -c
   1 /usr/local/bin/git-lfs prune
  47 git -c filter.lfs.smudge= -c filter.lfs.clean= -c filter.lfs.process= -c filter.lfs.required=false cat-file --batch-check
  65 git -c filter.lfs.smudge= -c filter.lfs.clean= -c filter.lfs.process= -c filter.lfs.required=false rev-list --stdin --objects --no-walk --
   1 git lfs prune
   1 grep git
```

There are no error messages to the console, and `run_command` doesn't seem to fail:

```
12:27 $ GIT_TRACE=1 git lfs prune 2>&1 | grep run_command
12:27:51.879582 run-command.c:643       trace: run_command: git-lfs prune
12:27:51.927361 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.096859 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.098864 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.100725 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.100752 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.100848 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.102625 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.108344 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.108923 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.114106 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.096778 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.123233 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.127792 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.128629 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.128834 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.129020 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.129218 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.129413 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.135609 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.127644 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.137172 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.137361 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.137425 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.117740 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.152899 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.145751 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.145817 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.145872 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.145904 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.169447 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.148367 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.169663 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.148553 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.172637 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.173596 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.173645 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.173982 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.179396 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.181824 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.182053 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.185531 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.186667 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.186775 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.191342 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.178278 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.198348 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.186284 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.187159 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.211047 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.224592 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.233195 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.237940 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.246638 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.246966 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.250340 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.250400 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.249924 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.250036 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.250232 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.292630 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.298152 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.300480 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.302367 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.302923 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.302962 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.303253 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.304354 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.312135 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.314990 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.310116 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.311425 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.311771 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.320211 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.311219 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.310287 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.310477 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.311621 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.311966 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.310684 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.310863 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.311025 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.325683 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.326400 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.326790 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.327123 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.335204 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.335422 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.342640 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.348342 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.352581 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.352641 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.352918 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.352974 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.353522 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.353907 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.354152 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.354396 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.354647 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.354904 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.355155 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.355752 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.357335 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.365233 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.367492 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.369284 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.376712 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.376741 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.376848 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.376931 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.377058 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.377115 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.377177 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.377217 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.377304 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.377358 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.377884 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.378627 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.379115 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.379323 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.379479 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.379986 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.384330 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.384414 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.384557 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.384731 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.384957 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.385188 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.385435 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.385682 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.384528 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.386179 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.389760 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.402318 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.410528 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.410618 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.410705 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.410808 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.410884 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.411088 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.411179 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
12:27:52.432284 trace git-lfs: run_command: git rev-list --stdin --objects --no-walk --
```

I cannot provide a reproduction outside of my checkout, but I can run tests against my checkout if there are any tests that could be run for more information.

**Comment by bk2204** [comment]
Hey, sorry to hear you're having trouble.

Did this issue occur when using a version between 2.5.0 and 2.5.2?  If not, can you try the latest `master` or `release-next` branch (the latter of which has binaries in #3339) and see if that fixes the problem?  We just fixed a hang that occurred in 2.6.0, and it would be helpful to know if our fix there also fixes your issue.

**Comment by pwagland** [comment]
I assume that should have been #3399.

Using:
```
17:05 $ git lfs env
git-lfs/2.6.1 (GitHub; darwin amd64; go 1.11.2; git 91a5ce40)
git version 2.19.2
```

I still get the hang, with roughly the above symptoms, I have slightly fewer child processes now:
```
17:11 $ ps -ef | grep git | cut -c 55- | sort | uniq -c
   1 /usr/local/bin/git-lfs prune
  43 git -c filter.lfs.smudge= -c filter.lfs.clean= -c filter.lfs.process= -c filter.lfs.required=false cat-file --batch-check
  64 git -c filter.lfs.smudge= -c filter.lfs.clean= -c filter.lfs.process= -c filter.lfs.required=false rev-list --stdin --objects --no-walk --
   1 git lfs prune
   1 grep git
```

**Comment by bk2204** [comment]
> I assume that should have been #3399.

Yeah, sorry about that.

It would be really helpful if you could attach the entire output from GIT_TRACE against the 2.6.1 prerelease, so I can see what items we're traversing and what items we're not.  That would let me identify where we're failing to see if I can narrow down the problem.

**Comment by pwagland** [comment]
I had a look at the logs, and since this is a private repo, there is probably too much information in there for me to share comfortably. So I decided to git bisect to see when the issue was introduced, and I noticed that on this repo, a lot of the old revisions fail with (for example from 1b1e1af8):

```
16:00 $ cat /Users/pwagland/work/src/main/.git/lfs/logs/20181205T160017.720803.log
git-lfs/2.5.0 (GitHub; darwin amd64; go 1.11.2)
git version 2.19.2

$ git-lfs prune
Prune error: fork/exec /usr/local/bin/git: too many open files
Prune error: open stderr: pipe: too many open files
Prune error: fork/exec /usr/local/bin/git: too many open files
Prune error: fork/exec /usr/local/bin/git: too many open files
```

**Comment by pwagland** [comment]
@bk2204 If I can provide the logs and hide all file/branch names, and this is still useful, then I can probably add those obfuscated logs to this issue.

**Comment by bk2204** [comment]
Yeah, that would be fine.

**Comment by juhanisimola-yousician** [comment]
I have the same problem, also with a private repository. We have 152 remote branches. When I delete the local references to all remote branches, `git lfs prune` works. When I pull the references are added back, and `git lfs prune` hangs again.

**Comment by bk2204** [comment]
I tried to reproduce this issue with a similar setup and I haven't been able to. A testcase of some sort would be really helpful for us in trying to track down this issue.

**Comment by romoore** [comment]
I'm having the same issue with https://github.com/material-components/material-components-ios . I've been working locally with a bunch of snapshot tests.

Now I'm on a brand new branch where I wrote a simple bash script.  I can't seem to push the new branch directly to the upstream (main) repo (not my clone) with failures from git lfs. When I try to prune, it gets stuck.  Can you try with material-components-ios?

**Comment by bk2204** [comment]
I've tried with that repo, and I'm still not seeing this. Could you provide output with `GIT_TRACE=1` so I can see what's going on when this occurs?

**Comment by romoore** [comment]
Here's the log. It ran at 100% for about 5 minutes and then just sort of stopped.

[git-lfs-prune.log](https://github.com/git-lfs/git-lfs/files/2703562/git-lfs-prune.log)

**Comment by bk2204** [comment]
I've put a branch up called `prune-hang` in my remote. Is it possible that you could try it and see if it works or at least produces a useful error?

**Comment by pwagland** [comment]
Sorry I am not participating here at the moment, it appears that my particular problem has gone away :-(, and thus I cannot reproduce with the old build, and thus also not verify that the new build is any better.

**Comment by pwagland** [comment]
> I couldn't reproduce the problem either with the material-components-ios. I wrote a little script that creates 150 branches and bingo, the problem appears.

I can confirm that I couldn't reproduce before this script, and after running this script I _could_ reproduce. Thanks @osimola!

**Comment by pwagland** [comment]
> I've put a branch up called `prune-hang` in my remote. Is it possible that you could try it and see if it works or at least produces a useful error?

As mentioned in my previous reply, this neither works, not produces an error. It just hangs, the same as the release version of `git-lfs`.

**Comment by bk2204** [comment]
Thanks for the testcase. I wasn't able to reproduce this with 150 or 250 items, but 550 seems to have done the trick.

**Comment by bk2204** [comment]
I've pushed a new `prune-hang` branch up. I believe the last branch was missing a line, which is why it didn't show the problem. I've included that fix and a fix for the underlying issue I saw as well, so hopefully things should be working now.

Please let me know how it works for you.

**Comment by pwagland** [comment]
Hi @bk2204 ,

It's not working for me :-(

Just to ensure that I am testing the right thing, this is what I did to test the fix:
```bash
10:39 $ rm -rf ~/go/
✔ ~/src/material-components-ios [branch_149 L|✔]
10:40 $ go get github.com/bk2204/git-lfs
✔ ~/src/material-components-ios [branch_149 L|✔]
10:41 $ (cd ~/go/src/github.com/bk2204/git-lfs; git checkout prune-hang)
Branch 'prune-hang' set up to track remote branch 'prune-hang' from 'origin'.
Switched to a new branch 'prune-hang'
✔ ~/src/material-components-ios [branch_149 L|✔]
10:41 $ go install github.com/bk2204/git-lfs
✔ ~/src/material-components-ios [branch_149 L|…1]
10:41 $ ~/go/bin/git-lfs prune
prune: 1354 local object(s), 1544 retained
```

At this point it still hangs for me. If this is the correct reproduction, then I will also attach the trace file.

**Comment by bk2204** [comment]
I'm not sure that `go install` is the best option for this situation. I personally just run `make` and symlink the binary into `PATH`. I think if you use `go install` here, it may pull down portions of the upstream repo that don't contain the proper changes, although I can't be certain.

So maybe something like this:

```shellSession
$ (cd ~/go/src/github.com/bk2204/git-lfs && git checkout prune-hang && make)
$ ~/go/src/github.com/bk2204/git-lfs/bin/git-lfs prune
```

**Comment by pwagland** [comment]
Hi @bk2204,

OK, this is what I did:

```bash
✔ ~/src/material-components-ios [branch_149 L|…3]
16:05 $ (cd ~/go/src/github.com/bk2204/git-lfs && git pull && git checkout prune-hang && make)
Already up to date.
Current branch prune-hang is up to date.
Already on 'prune-hang'
Your branch is up to date with 'origin/prune-hang'.
git-lfs: skipping fmt, no goimports found at `goimports` ...
go generate github.com/git-lfs/git-lfs/commands
GOOS= GOARCH= go build -ldflags=" -X github.com/git-lfs/git-lfs/config.GitCommit=289522be -s -w " -gcflags=" " -o ./bin/git-lfs ./git-lfs.go
✔ ~/src/material-components-ios [branch_149 L|…3]
16:05 $ GIT_TRACE=1 ~/go/src/github.com/bk2204/git-lfs/bin/git-lfs prune > git-lfs.trace.txt 2>&1
^C✘-INT ~/src/material-components-ios [branch_149 L|…3]
```

And this is the resulting [git-lfs trace file](https://github.com/git-lfs/git-lfs/files/2727434/git-lfs.trace.txt)… I killed it after a minute, all processes were hanging on 0% CPU usage.

**Comment by bk2204** [comment]
The problem I fixed in an earlier commit in that branch (fd52421b9) was that if the new object scanner can't open a file and returns an error, we wait on an error channel that will never be closed. I'm not sure why that fix isn't working for you, since we should be producing an error after that commit if the file cannot be opened.

I've pushed up a new commit that limits the number of ref scan threads to twice your CPUs instead of 100. On my 8-core laptop, that means it should require 160 file descriptors or fewer, which should be sensible for most systems (and hopefully for your system as well). I've confirmed that with the old code, we definitely ran out of file descriptors when running under `ulimit -n 160`.

Note that it will take a few seconds for things to start, since we'll be spawning a decent number of processes. Also, just to note, if you're building within the GOPATH, you'll probably want to make the name of the directory `github.com/git-lfs/git-lfs`, since Go will look for Git LFS's subpackages in that directory. I'm worried that we're pulling in a copy of the upstream repo that doesn't contain the fixes when things get built.

Also, just to satisfy my curiosity, what is the value for `ulimit -n`? It might be useful for us to set that value when running the testsuite if users are running with lower values in production.
